added ESM support

MarkusDeutschmann opened the pull request:
- added ESM support - `dist-esm` will be created in addition to dist
- removed `lodash` dependency for easier use of ESM
- because of ESM the static imports needed to be changed - now ending with *.js
- needed to replace `ts-prune` with `unimported` as `ts-prune` does not supporting ESM imports 
- updated dependencies

patik commented on module/src/common/options.ts:
```diff
@@ -74,11 +73,11 @@ export function getConfig(
         })
     } else {
-        settings = defaults({}, userOptions, defaultSettings)
+        settings = Object.assign({}, defaultSettings, userOptions);
```
We could use `lodash-es` and continue using `defaults()`, no? I believe I used it because it also copies nested properties.

MarkusDeutschmann replied:
I was just thinking get rid of lodash as a real dependency would be good as lodash (because it provides a lot of functionality) very often has issues which will show up here but as only defaults() is used it is not a real issue - but confusing. On the other hand I really don't see any difference between the both commands as they are here.
We only speak about object, right?
I compared them in a small test.js script: 

```
import defaults from "lodash-es/defaults.js";
const def = {
	name: 'test',
	age: 18,
	body: {
		height: 180,
		weight: 70
	},
	hobby: ['basketball', 'football'],
	devices: {
		phone: {
			brand: 'iphone',
			model: 'iphone 11'
		},
		computer: {
			brand: 'macbook',
			model: 'macbook pro 2019'
		}
	}
};
const override = {
	hobby: ['gaming'],
	body: {
		height: 190
	}, devices: {
		phone: {
			model: 'iphone 10 SE'
		}
	}
}

console.log("default object def:");
console.dir(def, { depth: 5 });
console.log("object that overwrite defaults:");
console.dir(override, { depth: 5 });
console.log("Object.assign:");
console.dir(Object.assign({}, def, override), { depth: 5 });
console.log("defaults:");
console.dir(defaults({}, override, def), { depth: 5 });
```
The result is: 
```
default object def:
{
  name: 'test',
  age: 18,
  body: { height: 180, weight: 70 },
  hobby: [ 'basketball', 'football' ],
  devices: {
    phone: { brand: 'iphone', model: 'iphone 11' },
    computer: { brand: 'macbook', model: 'macbook pro 2019' }
  }
}
object that overwrite defaults:
{
  hobby: [ 'gaming' ],
  body: { height: 190 },
  devices: { phone: { model: 'iphone 10 SE' } }
}
Object.assign:
{
  name: 'test',
  age: 18,
  body: { height: 190 },
  hobby: [ 'gaming' ],
  devices: { phone: { model: 'iphone 10 SE' } }
}
defaults:
{
  hobby: [ 'gaming' ],
  body: { height: 190 },
  devices: { phone: { model: 'iphone 10 SE' } },
  name: 'test',
  age: 18
}
```

If a more deeper merge is needed it could be done like that:
```
import defaults from "lodash-es/defaults.js";
const def = {
	name: 'test',
	age: 18,
	body: {
		height: 180,
		weight: 70
	},
	hobby: ['basketball', 'football'],
	devices: {
		phone: {
			brand: 'iphone',
			model: 'iphone 11'
		},
		computer: {
			brand: 'macbook',
			model: 'macbook pro 2019'
		}
	}
};
const override = {
	hobby: ['gaming'],
	body: {
		height: 190
	}, devices: {
		phone: {
			model: 'iphone 10 SE'
		}
	}
}

function isObject(item) {
	return (item && typeof item === 'object' && !Array.isArray(item));
}

function mergeDeep(target, ...sources) {
	if (!sources.length) {
		return target;
	}
	const source = sources.shift();

	if (isObject(target) && isObject(source)) {
		for (const key in source) {
		if (isObject(source[key])) {
			if (!target[key]) Object.assign(target, { [key]: {} });
			mergeDeep(target[key], source[key]);
		} else {
			Object.assign(target, { [key]: source[key] });
		}
		}
	}
	return mergeDeep(target, ...sources);
}

console.log("mergeDeep:");
console.dir(mergeDeep({}, def, override), { depth: 5 });
```
With that deeper merge the result would be:
```
default object def:
{
  name: 'test',
  age: 18,
  body: { height: 180, weight: 70 },
  hobby: [ 'basketball', 'football' ],
  devices: {
    phone: { brand: 'iphone', model: 'iphone 11' },
    computer: { brand: 'macbook', model: 'macbook pro 2019' }
  }
}
object that overwrite defaults:
{
  hobby: [ 'gaming' ],
  body: { height: 190 },
  devices: { phone: { model: 'iphone 10 SE' } }
}
mergeDeep:
{
  name: 'test',
  age: 18,
  body: { height: 190, weight: 70 },
  hobby: [ 'gaming' ],
  devices: {
    phone: { brand: 'iphone', model: 'iphone 10 SE' },
    computer: { brand: 'macbook', model: 'macbook pro 2019' }
  }
}
```
What you think about this?

MarkusDeutschmann commented on module/src/async/index.ts:
```diff
@@ -1,6 +1,6 @@
-import { getConfig } from '../common/options'
-import { MultipleSides, Side, UserOptions } from '../common/types'
-import { getIntersectionObserverOptions } from './options'
+import { getConfig } from '../common/options.js'
```
For ESM modules it is mandatory to import it with the extension the files will have after TypeScript compiled it. TypeScript itself won't do that.
It is described here: [ECMAScript Modules in Node.js](https://www.typescriptlang.org/docs/handbook/esm-node.html). Luckily CommonJS modules are also working with these imports.

MarkusDeutschmann commented:
Thanks. 
Here are the result of the test:
- yarn lint:
```
yarn run v1.22.19
$ yarn workspaces run lint

> withinviewport
$ yarn typecheck && yarn run code-lint && yarn unused-exports
$ tsc --noEmit
$ eslint --ext .ts --fix ./src/**/*.ts
$ unimported --no-cache
✓ There don't seem to be any unimported files.

> @patik/within-viewport-demo
$ yarn typecheck && yarn code-lint && yarn unused-exports
$ tsc --noEmit
$ eslint --ext .ts --fix ./src/**/*.ts
$ unimported --no-cache
✓ There don't seem to be any unimported files.
Done in 10.84s.
```
- yarn test:
```
yarn run v1.22.19
$ yarn --cwd module test
$ yarn jest --coverage
$ C:\git_repo\within-viewport\node_modules\.bin\jest --coverage
 PASS  src/tests/specs/strictCallbackAndRootMargin.test.ts
 PASS  src/tests/specs/options.test.ts
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   96.66 |    89.65 |     100 |   96.55 |                   
 options.ts |   95.45 |       88 |     100 |   95.45 | 47                
 sides.ts   |     100 |      100 |     100 |     100 |                   
------------|---------|----------|---------|---------|-------------------

Test Suites: 2 passed, 2 total  
Tests:       37 passed, 37 total
Snapshots:   0 total
Time:        3.21 s
Ran all test suites.
Done in 6.74s.
```
- yarn cover
```
yarn run v1.22.19
$ yarn --cwd module test --collectCoverage
$ yarn jest --coverage --collectCoverage
$ C:\git_repo\within-viewport\node_modules\.bin\jest --coverage --collectCoverage
 PASS  src/tests/specs/strictCallbackAndRootMargin.test.ts
 PASS  src/tests/specs/options.test.ts
------------|---------|----------|---------|---------|-------------------
File        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
------------|---------|----------|---------|---------|-------------------
All files   |   96.66 |    89.65 |     100 |   96.55 |                   
 options.ts |   95.45 |       88 |     100 |   95.45 | 47                
 sides.ts   |     100 |      100 |     100 |     100 |                   
------------|---------|----------|---------|---------|-------------------

Test Suites: 2 passed, 2 total  
Tests:       37 passed, 37 total
Snapshots:   0 total
Time:        2.625 s
Ran all test suites.
Done in 5.72s.
```
But no hurry regarding he release. If you want to check by yourself I understand that. It no issue to wait a little.

patik commented:
Thanks a lot. I will wait then, just in case something goes awry.

patik commented:
There were some conflicts with the yarn.lock that needed to be fixed, so I hope it's okay I pushed to your branch.

So far it looks good in terms of passing tests, etc. However I noticed an issue with the tree-shaking. I added this module to another project, imported just `{ withinviewport }`, and ran the Webpack bundle analyzer. It shows that the jQuery plugin is still being included:

<img width="737" alt="CleanShot 2023-06-26 at 22 33 49@2x" src="https://github.com/patik/within-viewport/assets/262137/1074727b-2035-4582-8131-71de9a5559b5">

Likely this is an issue with the way things are exported from index.ts and is not a problem with your PR. I'll need to look into reconfiguring the exports so that tree-shaking happens properly.

patik commented:
@MarkusDeutschmann I managed to get it working so that both CJS and ESM are in the same `dist` folder and tree-shaking works fine (e.g. when I use this package in another app, the jQuery plugin doesn't get bundled). Please take a look at the [recent changes](https://github.com/patik/within-viewport/compare/c6e812d08d2bf3c0b69be2e66fd18fe2136718ef..ec798a9ae75a90708bbe4dff6cc232b1b5d4322d) and you can merge it when you're ready.

MarkusDeutschmann commented:
Hi @patik Thanks a lot. From my point of view it you change are fine. The only thing that I changed is the postbuild step. I just moved it from the bash script into the package.json script section. Reason is that I would like to prevent the break of local dev builds on Windows systems. I hope this is ok for you.
Changes now merged.

patik commented:
@MarkusDeutschmann Looks good Works well on the Mac too.

Thanks again for the PR!
